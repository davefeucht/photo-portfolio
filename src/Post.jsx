import React from "react";
import axios from "axios";
import PropTypes from "prop-types";

class Post extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      thumbnailUrl: "",
      fullImageUrl: "",
      showFullImage: false,
      inList: true,
      errorMsg: ""
    };
  }

  _getPostImage() {
    axios.get("http://" + this.props.site + "/wp-json/wp/v2/media/" + this.props.image + "/")
      .then(res => {
        let thumbnailImage = new Image();
        let fullImage = new Image();
        const thumbnailUrl = res.data.media_details.sizes.thumbnail.source_url;
        const fullImageUrl = res.data.media_details.sizes.full.source_url;
        this.setState({ thumbnailUrl });
        this.setState({ fullImageUrl });
        thumbnailImage.src = thumbnailUrl;
        fullImage.src = fullImageUrl;
      }, error => {
        const errorMsg = "Did not work: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  } 

  _toggleDisplay(event) {
    event.preventDefault();
    this.props.clickImage(this.props.id);
    this.setState({showFullImage: !this.state.showFullImage}); 
  }

  componentWillMount() {
    this._getPostImage();
  }

  render() {
    
    let imgTag = "";
    let title = "";

    if(this.props.context == "thumbnail") {
      title = <div className="post-title-multiple"><a href="#" onClick={this._toggleDisplay.bind(this)}>{this.props.title}</a></div>;
      if(this.state.showFullImage) {
        imgTag = <div className="post-image"><img src={this.state.fullImageUrl} /></div>;
      }
      else {
        imgTag = <div className="post-image"><img src={this.state.thumbnailUrl} /></div>;
      }
    }
    else if(this.props.context == "category-image") {
      title = "";
      imgTag = <div className="post-image"><img src={this.state.thumbnailUrl} /></div>;
    }
    else {
      title = <div className="post-title-single"><a href="#" onClick={this._toggleDisplay.bind(this)}>{this.props.title}</a></div>;
      imgTag = <div className="post-image"><img src={this.state.fullImageUrl} /></div>;
    }

    return(
      <div className="post">
        {title}
        {imgTag}         
      </div>
    );

  }  

}

Post.propTypes = {
  id: PropTypes.number,
  title: PropTypes.string,
  image: PropTypes.number,
  context: PropTypes.string,
  site: PropTypes.string,
  clickImage: PropTypes.func
};

export default Post;
