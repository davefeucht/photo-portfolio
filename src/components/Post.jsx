/****************
* Post component displays one individual Post
****************/

import React from "react";
import axios from "axios";
import PropTypes from "prop-types";
import FullPost from "./styledComponents/FullPost.jsx";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import {setPostUrls} from "../actions/actions.js";

class Post extends React.Component {
  static propTypes = {
    id: PropTypes.number,
    title: PropTypes.string,
    image: PropTypes.number,
    context: PropTypes.string,
    site: PropTypes.string,
    clickImage: PropTypes.func
  };

  _getPostImage() {
    const getPostImageURI = `https://${this.props.siteUrl}/wp-json/wp/v2/media/${this.props.image}`; 
    axios.get(getPostImageURI)
      .then(res => {
        let thumbnailImage = new Image();
        let fullImage = new Image();
        const thumbnailUrl = res.data.media_details.sizes.thumbnail.source_url;
        const fullImageUrl = res.data.media_details.sizes.full.source_url;
        this.props.setPostUrls({ thumbnail_image: thumbnailUrl, full_image: fullImageUrl });
        thumbnailImage.src = thumbnailUrl;
        fullImage.src = fullImageUrl;
      });
  } 

  _showAllPosts(event) {
    event.preventDefault();
    this.props.clickImage(this.props.id);
  }

  componentWillMount() {
    this._getPostImage();
  }

  render() {
    
    let divStyle = {backgroundImage: "url(" + this.props.fullImageUrl + ")", backgroundRepeat: "no-repeat", backgroundPosition: "center center", backgroundSize: "contain"};
    let classList = "post";

    return(
      <FullPost className={classList} style={divStyle} onClick={this._showAllPosts.bind(this)}>
      </FullPost>
    );
  }  
}

const mapStateToProps = state => {
  return {
    thumbnailImageUrl: state.applicationState.singlePostToShow.thumbnail_image,
    fullImageUrl: state.applicationState.singlePostToShow.full_image,
    siteUrl: state.applicationState.siteUrl,
    showAllPosts: state.visibilityFilter.showAllPosts
  };
}

const mapDispatchToProps = dispatch => {
  return {
    setPostUrls: bindActionCreators(setPostUrls, dispatch)
  }; 
}

export default connect(mapStateToProps, mapDispatchToProps)(Post);