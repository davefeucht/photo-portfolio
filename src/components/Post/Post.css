.post {
  position: absolute;
  width: 80%;
  height: 80%;
  cursor: pointer;
  grid-column: 1 / 6;
  grid-row: 1/3;
}

.post-background {
  position: absolute; 
  top: 0; 
  left: 0; 
  width: 100%;
  height: 100%; 
  background-color: rgba(0, 0, 0, 0.8); 
  z-index: 10;
}