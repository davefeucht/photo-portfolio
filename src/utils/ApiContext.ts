import { createContext } from 'react';

import { API } from './types';

export const ApiContext = createContext<API | null>(null);
