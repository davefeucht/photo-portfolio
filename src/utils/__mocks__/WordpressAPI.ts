const imageUrl = 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg';

const siteInfo = {
    siteName: 'Through a Pinhole',
    siteUrl: 'throughapinhole.com'
};

const categories = [
    {
        id: 23,
        count: 6,
        description: 'Category of other things',
        link: '',
        name: 'Some other category',
        slug: 'smocg',
        taxonomy: '',
        parent: 12,
        meta: {}
    },
    {
        id: 35,
        count: 3,
        description: 'Category of things',
        link: '',
        name: 'Some category',
        slug: 'smcg',
        taxonomy: '',
        parent: 23,
        meta: {}
    }
];

const posts = [
    {
        id: 887,
        link: '',
        modified: '',
        slug: '',
        type: '',
        featured_media: 142,
        categories: [220],
        title: {
            rendered: 'Some post'
        },
        content: {
            rendered: 'Some content',
            protected: false
        },
        thumbnail_image: 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg'
    },
    {
        id: 889,
        featured_media: 145,
        categories: [220],
        title: {
            rendered: 'Some other post'
        },
        thumbnail_image: 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg'
    }
];

const pages = [
    {
        id: 150,
        content: {
            rendered: 'Contact me!'
        },
        title: {
            rendered: 'Contact'
        }
    }
];

const tagNames = ['something', 'somethingelse'];

export default class WordpressAPI {
    siteUrl: string;

    constructor(siteUrl: string) {
        this.siteUrl = siteUrl;
    }

    async getPostThumbnail() {
        return Promise.resolve(imageUrl);
    }

    async getSiteInfo() {
        return Promise.resolve(siteInfo.siteName);
    }

    async getCategories() {
        return Promise.resolve(categories);
    }

    async getCategoryImage() {
        return Promise.resolve(imageUrl);
    }

    async getPosts() {
        return Promise.resolve(posts);
    }

    async getPost() {
        return Promise.resolve(posts[0]);
    }

    async getCategoryInfo() {
        return Promise.resolve(categories[0]);
    }

    async getPostImage() {
        return Promise.resolve(imageUrl);
    }

    async getTagNames() {
        return Promise.resolve(tagNames);
    }

    async getPages() {
        return Promise.resolve(pages);
    }

    async getPage() {
        return Promise.resolve(pages[0]);
    }
}
