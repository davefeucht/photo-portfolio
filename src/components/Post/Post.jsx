/****************
* Post component displays one individual Post
****************/

import React from "react";
import axios from "axios";
import { observer } from 'mobx-react';
import './Post.css';

const Post = observer(({ props }) => {
  const _getPostImage = () => {
    const getPostImageURI = `https://${props.siteUrl}/wp-json/wp/v2/media/${this.props.image}`; 
    axios.get(getPostImageURI)
      .then(res => {
        let thumbnailImage = new Image();
        let fullImage = new Image();
        const thumbnailUrl = res.data.media_details.sizes.thumbnail.source_url;
        const fullImageUrl = res.data.media_details.sizes.full.source_url;
        props.setFullImageUrl(fullImageUrl);
        thumbnailImage.src = thumbnailUrl;
        fullImage.src = fullImageUrl;
      });
  } 

  const _showAllPosts = (event) => {
    event.preventDefault();
    props.clickImage(this.props.id);
  }
    
 _getPostImage();

  let divStyle = {backgroundImage: "url(" + props.fullImageUrl + ")", backgroundRepeat: "no-repeat", backgroundPosition: "center center", backgroundSize: "contain"};
  let classList = "post";

  return(
    <div className={classList} style={divStyle} onClick={_showAllPosts.bind(this)}>
    </div>
  );
});

Post.displayName = 'Post';

export default Post;