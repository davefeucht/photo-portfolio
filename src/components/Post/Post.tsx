/** **************
* Post component displays one individual Post in a modal
*************** */

import './Post.css';

import { reaction } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect, useRef, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';

import { getPost, getPostImage, getTagNames } from '../../utils/Api';
import {
    getNextPost,
    getPreviousPost,
    setPostRect
} from '../../utils/PostHelper';
import {
    Post as PostState,
    ScreenInfo,
    SiteInfo,
    VisiblePost
} from '../../utils/types';
import PostFooter from '../PostFooter/PostFooter';
import PostImage from '../PostImage/PostImage';
import PostTitlebar from '../PostTitlebar/PostTitlebar';

const image = document.createElement('img');

interface PostProps {
    screenInfo: ScreenInfo,
    siteInfo: SiteInfo,
    visiblePost: VisiblePost,
    currentCategoryPosts: PostState[],
    clearVisiblePostTagNames: () => void,
    setCurrentPost: (a: PostState, b: string[], c: string) => void
}

const Post: React.FC<PostProps> = ({
    screenInfo,
    siteInfo,
    visiblePost,
    currentCategoryPosts,
    clearVisiblePostTagNames,
    setCurrentPost
}) => {
    const [imageHeight, setImageHeight] = useState<number>(0);
    const { categoryId = '-1', postId = '-1' } = useParams();
    const navigate = useNavigate();
    const imageRef = useRef<HTMLImageElement>(null);

    const closeModalHandler = () => {
        navigate(`/category/${categoryId}`);
    };

    const stopPropagation = (e: React.MouseEvent) => {
        e.stopPropagation();
    };

    const updateImage = () => {
        setPostRect(image, screenInfo.width, screenInfo.height, visiblePost);
    };

    image.onload = () => {
        updateImage();
        setImageHeight(imageRef.current?.clientHeight ?? 0);
    };

    useEffect(() => {
        let tagNames: string[] = [];
        let url = '';
        if (postId) {
            clearVisiblePostTagNames();
            getPost(parseInt(postId), siteInfo.siteUrl)
                .then((post: PostState) => {
                    return post;
                })
                .then((newPost: PostState) => {
                    getTagNames(newPost.tags, siteInfo.siteUrl)
                        .then(tags => {
                            tagNames = tags;
                        });
                    getPostImage(newPost.featured_media, siteInfo.siteUrl)
                        .then((imageUrl: string) => {
                            url = imageUrl;
                            setCurrentPost(newPost, tagNames, url);
                        });
                });
        }
    }, [postId]);

    useEffect(() => {
        if (visiblePost.fullImageUrl) {
            image.src = visiblePost.fullImageUrl;
        }

        if (imageRef.current) {
            imageRef.current.addEventListener('transitionend', updateImage, true);
        }

        const disposer = reaction(
            () => [screenInfo.width, screenInfo.height],
            () => updateImage()
        );

        updateImage();

        return () => {
            if (imageRef.current) {
                imageRef.current.removeEventListener('transitionend', updateImage);
            }
            disposer();
        };
    }, [visiblePost.fullImageUrl]);

    return (
        <div className="post-background" onClick={closeModalHandler}>
            <div className="post" onClick={stopPropagation}>
                <PostTitlebar postTitle={visiblePost.postTitle} />
                <PostImage
                    imageUrl={visiblePost.fullImageUrl}
                    imageHeight={imageHeight}
                    previousPost={getPreviousPost(parseInt(postId), currentCategoryPosts)}
                    nextPost={getNextPost(parseInt(postId), currentCategoryPosts)}
                    ref={imageRef}
                />
                <PostFooter tagNames={visiblePost.tagNames} />
            </div>
        </div>
    );
};

Post.displayName = 'Post';

export default observer(Post);
