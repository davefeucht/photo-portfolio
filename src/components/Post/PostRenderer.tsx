import PostFooter from 'components/PostFooter/PostFooter';
import PostImage from 'components/PostImage/PostImage';
import PostTitlebar from 'components/PostTitlebar/PostTitlebar';
import { reaction } from 'mobx';
import * as React from 'react';
import { useEffect, useRef, useState } from 'react';
import { setPostRect } from 'utils/PostHelper';
import { ScreenInfo, VisiblePost } from 'utils/types';

interface Props {
    onClose(): void;
    previousPost: number;
    nextPost: number;
    screenInfo: ScreenInfo;
    visiblePost: VisiblePost;
}

const image = document.createElement('img');

const PostRenderer: React.FC<Props> = ({
    onClose,
    previousPost,
    nextPost,
    screenInfo,
    visiblePost
}) => {
    const [imageHeight, setImageHeight] = useState<number>(0);
    const imageRef = useRef<HTMLImageElement>(null);

    const updateImage = () => {
        setPostRect(image, screenInfo.width, screenInfo.height, visiblePost);
    };

    image.onload = () => {
        updateImage();
        setImageHeight(imageRef.current?.clientHeight ?? 0);
    };

    useEffect(() => {
        if (visiblePost.fullImageUrl) {
            image.src = visiblePost.fullImageUrl;
        }

        if (imageRef.current) {
            imageRef.current.addEventListener('transitionend', updateImage, true);
        }

        const disposer = reaction(
            () => [screenInfo.width, screenInfo.height],
            () => updateImage()
        );

        updateImage();

        return () => {
            if (imageRef.current) {
                imageRef.current.removeEventListener('transitionend', updateImage);
            }
            disposer();
        };
    }, [visiblePost.fullImageUrl]);

    return (
        <div className="post-background" onClick={onClose}>
            <div className="post" onClick={event => event.stopPropagation()}>
                <PostTitlebar postTitle={visiblePost.postTitle} />
                <PostImage
                    imageUrl={visiblePost.fullImageUrl}
                    imageHeight={imageHeight}
                    previousPost={previousPost}
                    nextPost={nextPost}
                    ref={imageRef}
                />
                <PostFooter tagNames={visiblePost.tagNames} />
            </div>
        </div>
    );
};

export default PostRenderer;
