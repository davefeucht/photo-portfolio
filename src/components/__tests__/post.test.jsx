import { render } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

import StateStore from '../../StateStore/store';
import { ApiContext } from '../../utils/ApiContext';
import WordpressAPI from '../../utils/WordpressAPI';
import Post from '../Post/Post';

const screenInfo = {
    width: 500,
    height: 500
};

const siteInfo = {
    siteName: 'Through a Pinhole',
    siteUrl: 'throughapinhole.com'
};

const visiblePost = {
    postId: 150,
    postTitle: 'foo',
    tags: [0, 1],
    tagNames: ['moe', 'curly'],
    fullImageUrl: 'http://foo',
    width: 500,
    height: 500,
    featured_media: 0
};

jest.mock('../../utils/WordpressAPI');

test('Post displays', async () => {
    const store = new StateStore();
    const api = new WordpressAPI(store.siteInfo.siteUrl);
    const posts = await api.getPosts();
    let container;
    await act(async () => {
        container = render(
            <MemoryRouter initialEntries={['/post/150']}>
                <ApiContext.Provider value={api}>
                    <Routes>
                        <Route
                            path="post/:postId"
                            element={(
                                <Post
                                    screenInfo={screenInfo}
                                    siteInfo={siteInfo}
                                    visiblePost={visiblePost}
                                    currentCategoryPosts={posts}
                                    clearVisiblePostTagNames={() => {}}
                                    setCurrentPost={store.setCurrentPost}
                                />
                            )}
                        />
                    </Routes>
                </ApiContext.Provider>
            </MemoryRouter>
        ).container;
    });
    expect(container.firstChild).toMatchSnapshot();
});
