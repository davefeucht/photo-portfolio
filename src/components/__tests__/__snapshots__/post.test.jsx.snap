// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post displays 1`] = `
<div
  class="post-background"
>
  <div
    class="post"
  >
    <div
      class="post-titlebar"
    >
      <div
        class="title"
      >
        foo
      </div>
      <div
        class="close-button"
        role="button"
      >
        <a
          href="/category/undefined"
        >
          <div
            class="close-button-content"
          >
            <img
              height="30px"
              src="./assets/images/close-button.svg"
              width="30px"
            />
          </div>
        </a>
      </div>
    </div>
    <div
      class="post-image"
    >
      <img
        src="http://foo"
      />
    </div>
    <div
      class="post-footer"
    >
      <div
        class="labels"
      >
        foo, bob
      </div>
    </div>
  </div>
</div>
`;
